File: Backtracking/39_CombinationSum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []

        def dfs(i, cur, total):
            if total == target:
                res.append(cur.copy())
                return
            if i >= len(candidates) or total > target:
                return

            cur.append(candidates[i])
            dfs(i, cur, total + candidates[i])

            cur.pop()
            dfs(i + 1, cur, total)

        dfs(0, [], 0)
        return res


class Solution2:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []

        def dfs(idx, cur, total):
            if total == target:
                res.append(cur.copy())
                return
            if idx >= len(candidates) or total > target:
                return
            for i in range(idx, len(candidates)):
                dfs(i, cur + [candidates[i]], total + candidates[i])

        dfs(0, [], 0)
        return res

File: Java Solutions/39_Combination Sum.java
class Solution {
    protected void backtrack(
        int remain,
        LinkedList<Integer> comb,
        int i,
        int[] candidates,
        List<List<Integer>> res
    ) {
        if (remain == 0) {
            res.add(new ArrayList<Integer>(comb));
            return;
        } else if (i >= candidates.length || remain < 0) {
            return;
        }

        comb.add(candidates[i]);
        backtrack(remain - candidates[i], comb, i, candidates, res);

        comb.removeLast();
        backtrack(remain, comb, i + 1, candidates, res);
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        LinkedList<Integer> comb = new LinkedList<Integer>();
        backtrack(target, comb, 0, candidates, res);
        return res;
    }
}